### ISSUE TITLE
Connection: close

### ISSUE THREAD

**Issue by NebzHB** [Apollon77/alexa-remote#32]
Hello,

from time to time, i receive a Connection: close from Amazon :

```
HTTP/1.1 0 x-amzn-RequestId: a33eaaaa-aaaa-aaaa-aaaa-8fdfdaaaab9e
Access-Control-Allow-Origin: https://alexa.amazon.fr
Vary: Origin
Access-Control-Expose-Headers: x-amzn-dat-gui-client-upgrade,X-Amzn-Error,loginUrl,X-Amzn-RequestId,x-amzn-alt-domain,Date,Location
Access-Control-Allow-Credentials: true
Date: Mon, 01 Apr 2019 15:41:37 GMT
Connection: close
```
of course the `ret = JSON.parse(body);` fail with that error : Response: No/Invalid JSON
(the exception is of course Unexpected token H in JSON at position 0)

do you know when that occurs ? and the solution ?

Thank you in advance

**Comment by Apollon77** [comment]
Which function/request is causing this?  May you send too fast?

**Comment by NebzHB** [comment]
i made a "daemon" that send Speak requests. after few hours without use, when i send a new speak request, i have that connection: close response... i've tried to checkAuthentication every 15min to keep the connection open but that doesn't work...

**Comment by Apollon77** [comment]
Can I see a full log and the script please?

**Comment by NebzHB** [comment]
you may it see there :
https://github.com/sigalou/jeedom_alexaapi/blob/master/resources/alexaapi.js

and the log (we've inserted code to detect the connexion: close):
```
[2019-4-3 22:50:16] Alexa-API: Alexa.Speak
[2019-4-3 22:50:16] Alexa-API: device: G090LF1181750XBN
[2019-4-3 22:50:16] Alexa-API: text: Ceci est un test de message pour la commande Speak
[2019-4-3 22:50:16] Alexa-sendSequenceCommand: else
[2019-4-3 22:50:16] Alexa-Remote: Sending Request with {"host":"alexa.amazon.fr","path":"/api/bootstrap?version=0","method":"GET","timeout":10000,"headers":{}}
[2019-4-3 22:50:17] Alexa-Remote: Response: {"authentication":{"authenticated":true,"canAccessPrimeMusicContent":true,"customerEmail":"****@free.fr","customerId":"A2****","customerName":"********"}}
[2019-4-3 22:50:17] Alexa-Remote: Authentication check successfull
[2019-4-3 22:50:17] Alexa-Remote: Sending Request with {"host":"alexa.amazon.fr","path":"/api/behaviors/preview","method":"POST","timeout":10000,"headers":{}}and data={"behaviorId":"PREVIEW","sequenceJson":"{\"@type\":\"com.amazon.alexa.behaviors.model.Sequence\",\"startNode\":{\"@type\":\"com.amazon.alexa.behaviors.model.OpaquePayloadOperationNode\",\"operationPayload\":{\"deviceType\":\"A3S5BH2HU6VAYF\",\"deviceSerialNumber\":\"****\",\"locale\":\"fr-FR\",\"customerId\":\"***\",\"textToSpeak\":\"Ceci est un test de message pour la commande Speak\"},\"type\":\"Alexa.Speak\"}}","status":"ENABLED"}
[2019-4-3 22:50:17] ******************************************************
[2019-4-3 22:50:17] *********************DEBUG****************************
[2019-4-3 22:50:17] ******************************************************
[2019-4-3 22:50:17] **DEBUG**DEBUG*Alexa-Remote: Response: No/Invalid JSON
[2019-4-3 22:50:17] HTTP/1.1 0 x-amzn-RequestId: 780ddb4a-565a-11e9-98d7-************
Access-Control-Allow-Origin: https://alexa.amazon.fr
Vary: Origin
Access-Control-Expose-Headers: x-amzn-dat-gui-client-upgrade,X-Amzn-Error,loginUrl,X-Amzn-RequestId,x-amzn-alt-domain,Date,Location
Access-Control-Allow-Credentials: true
Date: Wed, 03 Apr 2019 21:50:16 GMT
Connection: close
[2019-4-3 22:50:17] **DEBUG**DEBUG* Message Exception :Unexpected token H in JSON at position 0
[2019-4-3 22:50:17] ******************************************************
[2019-4-3 22:50:17] ******************************************************
[2019-4-3 22:50:17] ******************************************************
[2019-4-3 22:50:17] ***************FOUND**CONNEXION CLOSE *****************
[2019-4-3 22:50:17] ******************************************************
[2019-4-3 22:50:17] Alexa-API: ******************************************************************
[2019-4-3 22:50:17] Alexa-API: *****************************ERROR********************************
[2019-4-3 22:50:17] Alexa-API: ******************************************************************
[2019-4-3 22:50:17] Connexion Close
[2019-4-3 22:50:17] Alexa-API: ******************************************************************
Error: Connexion Close
at IncomingMessage.res.on (/var/www/html/plugins/alexaapi/resources/lib/alexa-remote.js:524:52)
at emitNone (events.js:111:20)
at IncomingMessage.emit (events.js:208:7)
at endReadableNT (_stream_readable.js:1064:12)
at _combinedTickCallback (internal/process/next_tick.js:139:11)
at process._tickCallback (internal/process/next_tick.js:181:9)
```

**Comment by Apollon77** [comment]
Hm ... I think the "Connection:Close" is not the topic ... the question is why in this case the whole HTTP header is given back as body ... this is the strange part. It more seems to be an invalid Amazon response whatI  never saw so far! So I have no idea about it. When you try to repeat is it then working?

**Comment by NebzHB** [comment]
yes with the normal code, we just receive NO/Invalid JSON response, that's why we did displayed the body...
Indeed a header in the body is strange... and the Header is not a valid one, normally the "0" after the HTTP/1.1 should be the HTTP return code (200, ...)

**Comment by Apollon77** [comment]
Yes you are exactly right and I think that's why the request is not fully parsed ... so something strange on amazon side it seems to me ... I never experienced this ...
I could only imagine that you are doing too many requests in too short time and run into their rate limiting ... Could this be?

**Comment by NebzHB** [comment]
sorry didn't reply to repeat : if we repeat after that it don't work, we have to startServer() again.

i don't think it's too many requests, we've tried to overload the requests and we get a proper : Alexa-Remote: Response: {"message":"Rate exceeded"}"

**Comment by Apollon77** [comment]
why that? Why you not simply make a second call if the first returns an error? YOu can check in the callback before calling "next" in your request handlng or such ...
One more thing: WHen you have multiple devices your code is sending out the request in parallel, maybe ewait till one is finished before sending the next one?

**Comment by NebzHB** [comment]
because if we make a second call, it doesn't work.

how do you know the first "Speak" is finished ?

**Comment by Apollon77** [comment]
> because if we make a second call, it doesn't work.

What error you get for a seond call?

> how do you know the first "Speak" is finished ?

In this case (multiple devices) it is just about sending the requests "queued" ... so when callbackis called.

**Comment by NebzHB** [comment]
1. i've asked to my users (i don't reproduce) (fyi it's a plugin for a domotic system)

2. the callback is called when the request is accepted, not when all the sentence is spoken (for the Speak example)

**Comment by Apollon77** [comment]
2.) Yes I know my note was also not about "Longer sentences to speak" it was just that you send out rewquests in parallel when sending to multiple devices.
There is no real easy way to etect when speak is over. YOu could only send a sequence of commands and maybe could control some virtual device afterwards so that you have a feedback but this is all määhh

**Comment by NebzHB** [comment]
2) ok i'll maybe "promise" it then... easier to queue

**Comment by NebzHB** [comment]
1. second call response is connexion: close too

**Comment by Apollon77** [comment]
Hm ... no idea ... never experienced this :-(

**Comment by NebzHB** [comment]
Thank you For your Help,I'll dig into these tracks.

As enhancement, That would be nice to promise-ing the commands to alexa, that would be easier for command chaining ;-)

**Comment by Apollon77** [comment]
Please open a separate issue as Feature request for it. And I‘m also happy to accept PRs ;-)
